### PR TITLE
Clean up dataset

### PR DESCRIPTION
**Pull request by cosmicBboy**
## Improve `dataset` methods API

- implement `dataset.loader` function: this is an optional function to handle cases where the output of `reader` is in some format that needs to be deserialized and loaded into memory
- implement `dataset.feature_loader`: similar to `loader` but this is used for deserializing features data or converting from a raw format on the prediction side of things.
- implement `dataset.feature_transformer`: a function applied to the output of `feature_loader`, for further data processing
- Add `feature_key` to `dataset.parser` so that user can indicate position of the features in the `parser` function output
- Updated docs.

### CODE REVIEW

File: unionml/fastapi.py [hunk 1]
```diff
@@ -49,16 +43,15 @@ def root():
     @app.post("/predict")
     async def predict(
         inputs: Optional[Union[Dict[str, Any], BaseModel]] = Body(None),
-        features: Optional[List[Dict[str, Any]]] = Body(None),
+        features: Optional[Union[List[Dict[str, Any]], str]] = Body(None),
```
Comment (eapolinario) [inline unionml/fastapi.py:46]: how hard would it be to add a test for this change?

File: unionml/dataset.py [hunk 2]
```diff
@@ -34,39 +37,69 @@ def __init__(
...
-        # TODO: make sure function signature matches expected signature:
-        # - splitter(data, test_size, shuffle, random_state)
+        """Register an optional splitter function that partitions data into training and test sets."""
         self._splitter = fn
         return fn
 
-    def parser(self, fn):
-        # TODO: make sure function signature matches expected signature:
-        # - parser(data, features, targets)
+    def parser(self, fn, feature_key: int = 0):
```
Comment (eapolinario) [inline unionml/dataset.py:74]: Should the type of `feature_key` be more restrictive? I'm concerned about discoverability.

**Review by eapolinario — COMMENTED** [review]
It's looking good, the only concern I have is with `feature_key` and how we're going to document / teach users about it.

File: unionml/fastapi.py (hunk 1 above)
Comment (cosmicBboy) [reply unionml/fastapi.py:46]: ah, I didn't mean to check this change in, going to revert

File: unionml/dataset.py (hunk 2 above)
Comment (cosmicBboy) [reply unionml/dataset.py:74]: what do you suggest?

**Pull request merged**